### PR TITLE
cuda: accelerate long-context B1 indexer and 1M HCA

### LINKED ISSUE #762: CUDA: B1 indexer misses WMMA and 1M HCA falls off the 7936-row fast path
## Summary

On current `main` (`84cc882`), DeepSeek-V4-Flash CUDA B1 decode has two
long-context cliffs on DGX Spark / GB10:

1. The ratio-4 CSA indexer always takes `indexer_score_one_direct_kernel` for
   `n_tokens == 1`, even when the existing WMMA128 scorer is faster.
2. At 1M context, ratio-128 HCA has 8192 compressed rows. The shared score cap is
   8192 total rows but reserves 256 raw rows, so only 7936 compressed rows fit.
   The last 256 compressed rows force all 20 HCA layers onto
   `attention_decode_mixed_heads8_online_kernel` instead of the existing exact
   score-split path.

Together these account for most of the 512K -> 1M decode collapse.

## Reproduction environment

- DS4: `84cc882`
- Hardware: DGX Spark, NVIDIA GB10 (`sm_121`)
- Model: DeepSeek-V4-Flash IQ2XXS / Q2_K / Q8 projections and output
- Model residency reported by DS4: 80.76 GiB
- `DS4_CUDA_Q8_F16_CACHE_MB=0`
- 1M session plan: 13.46 GiB KV + 0.06 GiB buffers + 80.76 GiB model =
  94.28 GiB

The benchmark allocates and touches the real CUDA F32-slot KV/indexer caches,
places the session at `context - 1`, and measures a complete target-only decode
token. It does not include a synthetic 1M prefill in TPOT. Legacy and optimized
paths alternate in the same process/session; each has 3 warmups and 30 samples.

## End-to-end A/B

| Context | Current TPOT | Patched TPOT | Current tok/s | Patched tok/s | Speedup |
|---:|---:|---:|---:|---:|---:|
| 128K | 82.182 ms | 80.208 ms | 12.168 | 12.468 | 1.025x |
| 256K | 93.963 ms | 88.571 ms | 10.642 | 11.290 | 1.061x |
| 512K | 115.649 ms | 104.450 ms | 8.647 | 9.574 | 1.107x |
| 1M | 228.729 ms | 143.250 ms | 4.372 | 6.981 | 1.597x |

## Kernel evidence

Microbenchmarks:

| Context | CSA direct | CSA WMMA | HCA current | HCA score-split |
|---:|---:|---:|---:|---:|
| 128K | 0.574 ms | 0.438 ms | 0.125 ms | 0.126 ms |
| 256K | 1.050 ms | 0.792 ms | 0.230 ms | 0.232 ms |
| 512K | 1.919 ms | 1.359 ms | 0.428 ms | 0.428 ms |
| 1M | 3.713 ms | 2.502 ms | 2.925 ms | 0.849 ms |

Nsight Systems capture of one complete 1M decode (capture excludes startup,
cache initialization, and warmup):

### Current

- `attention_decode_mixed_heads8_online_kernel`: 97.286 ms / 20 instances
- `indexer_score_one_direct_kernel`: 78.036 ms / 21 instances
- Captured wall TPOT: 261.230 ms

### Patched

- `indexer_scores_wmma128_kernel`: 48.565 ms / 21 instances
- `attention_decode_score_split_finalize_kernel`: 16.158 ms / 22 instances
- `attention_decode_score_split_scores_tile512_kernel`: 9.242 ms / 22 instances
- Captured wall TPOT: 148.228 ms

The profiled wall times include Nsight overhead and are not the primary
throughput numbers, but the kernel attribution is unambiguous.

## Proposed fix

- Increase `DS4_CUDA_ATTENTION_SCORE_CAP` from 8192 to 8448. This is the exact
  8192 compressed + 256 raw capacity needed by the 1M HCA shape and keeps the
  already-existing exact score-split implementation active.
- For non-quality B1 indexer calls, use the existing WMMA path once
  `n_comp >= 8192`; keep the direct scorer for short contexts, quality mode,
  and `DS4_CUDA_NO_INDEXER_WMMA`.

## Correctness checks

- QAT-format B1 score/top-k at `n_comp=8192`: max score delta
  `6.10352e-05`, top-512 index differences `0/512`.
- Isolated HCA output at 8192 compressed rows: max delta `0` on the regression
  vector.
- Full synthetic-frontier logits at 128K/256K/512K: bit-identical.
- Full synthetic-frontier logits at 1M: max abs `0.0469046`, RMSE `0.00876944`,
  same argmax. This is the expected reduction-order change from the long online
  fallback to DS4's existing exact score-split path.
- Real 30,474-token story fact-recall regression: passes.
- `make cuda-regression CUDA_ARCH=sm_121`: passes.

## Existing issue/PR audit

- #258 targets the same long-context indexer cost with approximate HISA
  pruning. It is based on an older revision, currently conflicts with `main`,
  and reports +1.9% at 256K. The change proposed here accelerates the current
  flat full-top-512 scorer and is complementary to a future HISA path.
- #586 is a Metal/SSD shared-event serialization problem.
- #416 and #418 are Metal packed-KV / prefill-memory work.
- #644 is about token-tiling prefill scratch, not B1 decode.
- #588 was benchmark tooling only and closed without merge.

I did not find an existing issue or PR for the CUDA HCA 7936-row capacity
fallback.

### PR DESCRIPTION
**Pull request by wangshang23**
Fixes #762.

## What changed

- Route non-quality B1 indexer scoring to the existing WMMA scorer at
  `n_comp >= 8192`; retain the direct scorer for shorter contexts, quality
  mode, and the existing `DS4_CUDA_NO_INDEXER_WMMA` escape hatch.
- Raise the attention score capacity from 8192 to 8448 rows so the 1M HCA
  shape fits 8192 compressed rows plus the existing 256-row raw reserve.
- Extend the CUDA long-context smoke test to cover 8192 compressed HCA rows
  and direct-vs-WMMA B1 score/top-k equivalence on QAT-format values.

The HCA change does not add a new attention implementation. It keeps the
existing exact score-split path active instead of falling back to the slow
online kernel for the last 256 rows at 1M.

## DGX Spark results

Base: `84cc882`, NVIDIA GB10 (`sm_121`), DeepSeek-V4-Flash IQ2XXS Q2 model,
Q8-to-F16 weight cache disabled. A/B paths alternate in one process/session,
with 3 warmups and 30 samples each.

| Context | Base tok/s | PR tok/s | Speedup |
|---:|---:|---:|---:|
| 128K | 12.168 | 12.468 | 1.025x |
| 256K | 10.642 | 11.290 | 1.061x |
| 512K | 8.647 | 9.574 | 1.107x |
| 1M | 4.372 | 6.981 | 1.597x |

At 1M, Nsight Systems attributes the main change to:

- CSA: direct scorer `78.036 ms` -> WMMA scorer `48.565 ms` across 21 layers.
- HCA: online kernel `97.286 ms` -> exact score-split score+finalize
  `25.400 ms` across 20 layers.

Full benchmark details and the existing-work audit are in #762. Related #258
uses approximate HISA pruning and is complementary to this full-top-512 flat
scorer optimization.

## Correctness

- B1 QAT score max delta: `6.10352e-05`.
- B1 top-512 differences: `0/512`.
- Isolated 8192-row HCA regression output delta: `0`.
- Full synthetic-frontier logits are bit-identical through 512K.
- At 1M, legacy online vs existing exact score-split has max logit delta
  `0.0469046`, RMSE `0.00876944`, and identical argmax.
- Real 30,474-token story fact-recall regression passes.

## Validation

```sh
make cuda-regression CUDA_ARCH=sm_121
```

Passes on DGX Spark / GB10.

`--logprob-vectors` currently fails `short_code_completion` identically on
both base and PR builds with this GGUF, so it is not a regression from this
change.

### CODE REVIEW

**Comment by JordiPosthumus** [comment]
Additional independent GB10 data point, tested against the current upstream base.

Exact sources:

- base: `b0a147a7fba6d1a104d047d5a140e9bb4bfc13cd`
- this PR: `fcca7b56d0eb2ee7592f8f10ece0efef3cb8553c`
- NVIDIA GB10 (`sm_121`), DeepSeek V4 Flash Vision Exp IQ2XXS/Q2
- 262,144-token allocation, 2,048-token prefill chunks, 64 generated tokens per frontier

The four arms ran serially with no competing DS4 process. This is one pass rather than a confidence interval, but the long-context result agrees closely with the PR's reported 256K gain:

| Context | Base prefill | PR prefill | Base steady decode | PR steady decode |
|---:|---:|---:|---:|---:|
| 65,536 | 806.93 | 810.97 | 13.58 | 13.53 |
| 131,072 | 767.68 | 771.78 | 12.33 | 12.56 (+1.9%) |
| 260,096 | 610.10 | 613.63 | 10.70 | 11.28 (+5.4%) |

Correctness evidence:

- `make cuda-regression CUDA_ARCH=sm_121` passed.
- B1 direct-vs-WMMA check: max absolute score delta `6.10352e-05`, `0` top-k differences.
- Full-vocabulary frontier logits were dumped at 2K, 4K, 8K, 16K, 32K, 65K, 131K, and 260K. Every SHA-256 matched the corresponding baseline dump exactly.

So on this second GB10/model configuration, #763 is neutral below the threshold and gives the expected material long-context decode gain without changing the output vectors.

The tested integration commit is [`f1218cc`](https://github.com/JordiPosthumus/ds4/commit/f1218cc) (followed only by common build/test prerequisites in the evaluation branch).
